File: app/python/src/grokking_fp/ch10_concurrency.py
"""第10章: 並行・並列処理

Python で並行処理を扱う関数型プログラミングのパターンを学びます。
asyncio を使った非同期処理、Ref（アトミック参照）の実装を学びます。
"""

import asyncio
import random
import threading
from collections.abc import Callable, Coroutine
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

T = TypeVar("T")
U = TypeVar("U")


# =============================================================================
# 10.1 Ref - アトミックな共有状態
# =============================================================================


class Ref(Generic[T]):
    """スレッドセーフなアトミック参照。

    複数の並行処理から安全にアクセスできる共有状態を提供します。

    Examples:
        >>> ref = Ref(0)
        >>> ref.get()
        0
        >>> ref.update(lambda x: x + 1)
        >>> ref.get()
        1
    """

    def __init__(self, initial: T) -> None:
        self._value = initial
        self._lock = threading.Lock()

    def get(self) -> T:
        """現在の値を取得する。"""
        with self._lock:
            return self._value

    def set(self, value: T) -> None:
        """値を設定する。"""
        with self._lock:
            self._value = value

    def update(self, f: Callable[[T], T]) -> None:
        """アトミックに値を更新する。"""
        with self._lock:
            self._value = f(self._value)

    def modify(self, f: Callable[[T], tuple[T, U]]) -> U:
        """アトミックに値を更新し、結果を返す。"""
        with self._lock:
            new_value, result = f(self._value)
            self._value = new_value
            return result

    def get_and_set(self, value: T) -> T:
        """値を設定し、古い値を返す。"""
        with self._lock:
            old_value = self._value
            self._value = value
            return old_value


class AsyncRef(Generic[T]):
    """非同期コンテキスト用のアトミック参照。

    asyncio との組み合わせで使用します。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     ref = AsyncRef(0)
        ...     await ref.update(lambda x: x + 1)
        ...     return await ref.get()
        >>> asyncio.run(example())
        1
    """

    def __init__(self, initial: T) -> None:
        self._value = initial
        self._lock = asyncio.Lock()

    async def get(self) -> T:
        """現在の値を取得する。"""
        async with self._lock:
            return self._value

    async def set(self, value: T) -> None:
        """値を設定する。"""
        async with self._lock:
            self._value = value

    async def update(self, f: Callable[[T], T]) -> None:
        """アトミックに値を更新する。"""
        async with self._lock:
            self._value = f(self._value)

    async def modify(self, f: Callable[[T], tuple[T, U]]) -> U:
        """アトミックに値を更新し、結果を返す。"""
        async with self._lock:
            new_value, result = f(self._value)
            self._value = new_value
            return result


# =============================================================================
# 10.2 チェックインのリアルタイム集計
# =============================================================================


@dataclass(frozen=True)
class City:
    """都市を表すデータクラス。"""

    name: str


@dataclass(frozen=True)
class CityStats:
    """都市のチェックイン統計。"""

    city: City
    check_ins: int


def top_cities(city_check_ins: dict[City, int], n: int = 3) -> list[CityStats]:
    """チェックイン数上位の都市を取得する（純粋関数）。

    Examples:
        >>> cities = {City("Tokyo"): 100, City("Osaka"): 50, City("Kyoto"): 75}
        >>> result = top_cities(cities, 2)
        >>> [s.city.name for s in result]
        ['Tokyo', 'Kyoto']
    """
    stats = [CityStats(city, count) for city, count in city_check_ins.items()]
    return sorted(stats, key=lambda s: s.check_ins, reverse=True)[:n]


def update_check_ins(
    city_check_ins: dict[City, int], city: City
) -> dict[City, int]:
    """チェックインを追加する（純粋関数）。

    Examples:
        >>> cities = {City("Tokyo"): 10}
        >>> result = update_check_ins(cities, City("Tokyo"))
        >>> result[City("Tokyo")]
        11
    """
    new_check_ins = dict(city_check_ins)
    new_check_ins[city] = new_check_ins.get(city, 0) + 1
    return new_check_ins


# =============================================================================
# 10.3 並行処理 - asyncio
# =============================================================================


async def async_sleep(seconds: float) -> None:
    """非同期スリープ（スレッドをブロックしない）。

    Examples:
        >>> import asyncio
        >>> asyncio.run(async_sleep(0.01))
    """
    await asyncio.sleep(seconds)


async def par_sequence(
    tasks: list[Coroutine[Any, Any, T]],
) -> list[T]:
    """複数の非同期タスクを並列実行する。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     async def task(n):
        ...         return n * 2
        ...     return await par_sequence([task(1), task(2), task(3)])
        >>> asyncio.run(example())
        [2, 4, 6]
    """
    return list(await asyncio.gather(*tasks))


async def par_traverse(
    items: list[T],
    f: Callable[[T], Coroutine[Any, Any, U]],
) -> list[U]:
    """リストの各要素に非同期関数を並列適用する。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     async def double(n):
        ...         return n * 2
        ...     return await par_traverse([1, 2, 3], double)
        >>> asyncio.run(example())
        [2, 4, 6]
    """
    return await par_sequence([f(item) for item in items])


# =============================================================================
# 10.4 サイコロを並行して振る
# =============================================================================


def cast_the_die_impure() -> int:
    """不純な関数: サイコロを振る。

    Examples:
        >>> result = cast_the_die_impure()
        >>> 1 <= result <= 6
        True
    """
    return random.randint(1, 6)


async def cast_the_die() -> int:
    """非同期でサイコロを振る。

    Examples:
        >>> import asyncio
        >>> result = asyncio.run(cast_the_die())
        >>> 1 <= result <= 6
        True
    """
    await asyncio.sleep(0)  # yield control
    return cast_the_die_impure()


async def cast_the_die_twice() -> int:
    """2つのサイコロを並行して振り、合計を返す。

    Examples:
        >>> import asyncio
        >>> result = asyncio.run(cast_the_die_twice())
        >>> 2 <= result <= 12
        True
    """
    results = await par_sequence([cast_the_die(), cast_the_die()])
    return sum(results)


async def cast_dice_parallel(n: int) -> list[int]:
    """n個のサイコロを並行して振る。

    Examples:
        >>> import asyncio
        >>> results = asyncio.run(cast_dice_parallel(3))
        >>> len(results)
        3
        >>> all(1 <= r <= 6 for r in results)
        True
    """
    return await par_sequence([cast_the_die() for _ in range(n)])


# =============================================================================
# 10.5 Ref を使った並行カウント
# =============================================================================


async def increment_three_times() -> int:
    """カウンターを3回インクリメントする。

    Examples:
        >>> import asyncio
        >>> asyncio.run(increment_three_times())
        3
    """
    counter = AsyncRef(0)
    await counter.update(lambda x: x + 1)
    await counter.update(lambda x: x + 1)
    await counter.update(lambda x: x + 1)
    return await counter.get()


async def sum_parallel(
    task1: Coroutine[Any, Any, int],
    task2: Coroutine[Any, Any, int],
    task3: Coroutine[Any, Any, int],
) -> int:
    """3つの非同期タスクを並列実行し、結果の合計を返す。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     async def pure(n):
        ...         return n
        ...     return await sum_parallel(pure(1), pure(2), pure(3))
        >>> asyncio.run(example())
        6
    """
    results = await par_sequence([task1, task2, task3])
    return sum(results)


async def count_evens(tasks: list[Coroutine[Any, Any, int]]) -> int:
    """複数の非同期タスクを並列実行し、偶数の結果をカウントする。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     async def pure(n):
        ...         return n
        ...     tasks = [pure(1), pure(2), pure(3), pure(4)]
        ...     return await count_evens(tasks)
        >>> asyncio.run(example())
        2
    """
    counter = AsyncRef(0)

    async def count_if_even(task: Coroutine[Any, Any, int]) -> None:
        result = await task
        if result % 2 == 0:
            await counter.update(lambda x: x + 1)

    await par_sequence([count_if_even(t) for t in tasks])
    return await counter.get()


# =============================================================================
# 10.6 チェックイン処理の並行版
# =============================================================================


async def store_check_in(
    stored_check_ins: AsyncRef[dict[City, int]], city: City
) -> None:
    """チェックインを保存する。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     ref = AsyncRef({})
        ...     await store_check_in(ref, City("Tokyo"))
        ...     return await ref.get()
        >>> asyncio.run(example())
        {City(name='Tokyo'): 1}
    """
    await stored_check_ins.update(lambda m: update_check_ins(m, city))


async def update_ranking(
    stored_check_ins: AsyncRef[dict[City, int]],
    stored_ranking: AsyncRef[list[CityStats]],
) -> None:
    """ランキングを更新する（1回）。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     check_ins = AsyncRef({City("Tokyo"): 10, City("Osaka"): 5})
        ...     ranking = AsyncRef([])
        ...     await update_ranking(check_ins, ranking)
        ...     result = await ranking.get()
        ...     return [s.city.name for s in result]
        >>> asyncio.run(example())
        ['Tokyo', 'Osaka']
    """
    check_ins = await stored_check_ins.get()
    await stored_ranking.set(top_cities(check_ins))


# =============================================================================
# 10.7 Fiber - 軽量タスク
# =============================================================================


@dataclass
class Fiber(Generic[T]):
    """軽量タスク（asyncio.Task のラッパー）。

    キャンセル可能なバックグラウンドタスクを表現します。
    """

    _task: asyncio.Task[T]

    def cancel(self) -> None:
        """タスクをキャンセルする。"""
        self._task.cancel()

    async def join(self) -> T:
        """タスクの完了を待つ。"""
        return await self._task

    @property
    def is_done(self) -> bool:
        """タスクが完了しているか。"""
        return self._task.done()


def start_fiber(coro: Coroutine[Any, Any, T]) -> Fiber[T]:
    """コルーチンを Fiber として起動する。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     async def task():
        ...         await asyncio.sleep(0.01)
        ...         return 42
        ...     fiber = start_fiber(task())
        ...     return await fiber.join()
        >>> asyncio.run(example())
        42
    """
    task = asyncio.create_task(coro)
    return Fiber(task)


async def forever(action: Callable[[], Coroutine[Any, Any, T]]) -> None:
    """アクションを永遠に繰り返す。

    Note:
        この関数は永遠に実行されるため、外部からキャンセルする必要があります。
    """
    while True:
        await action()


# =============================================================================
# 10.8 タイムアウト付き収集
# =============================================================================


async def collect_for(
    duration_seconds: float, interval_seconds: float = 0.1
) -> list[int]:
    """指定時間、定期的に乱数を収集する。

    Examples:
        >>> import asyncio
        >>> results = asyncio.run(collect_for(0.25, 0.1))
        >>> 1 <= len(results) <= 5  # 約2-3個
        True
    """
    collected: AsyncRef[list[int]] = AsyncRef([])

    async def producer() -> None:
        while True:
            await asyncio.sleep(interval_seconds)
            value = random.randint(0, 99)
            await collected.update(lambda lst: lst + [value])

    fiber = start_fiber(producer())

    await asyncio.sleep(duration_seconds)
    fiber.cancel()

    try:
        await fiber.join()
    except asyncio.CancelledError:
        pass

    return await collected.get()


# =============================================================================
# 10.9 並行マップ更新
# =============================================================================


@dataclass(frozen=True)
class Update:
    """マップ更新を表すデータクラス。"""

    key: str
    value: int


async def apply_updates(updates: list[Update]) -> dict[str, int]:
    """複数の更新を並行してマップに適用する。

    Examples:
        >>> import asyncio
        >>> updates = [Update("a", 1), Update("b", 2), Update("a", 3)]
        >>> result = asyncio.run(apply_updates(updates))
        >>> result["a"] in [1, 3]  # 並行実行のため順序は不定
        True
        >>> result["b"]
        2
    """
    map_ref: AsyncRef[dict[str, int]] = AsyncRef({})

    async def apply_update(update: Update) -> None:
        await map_ref.update(
            lambda m: {**m, update.key: update.value}
        )

    await par_sequence([apply_update(u) for u in updates])
    return await map_ref.get()


# =============================================================================
# 10.10 ProcessingCheckIns - バックグラウンド処理
# =============================================================================


@dataclass
class ProcessingCheckIns:
    """チェックイン処理の制御ハンドル。"""

    current_ranking: Callable[[], Coroutine[Any, Any, list[CityStats]]]
    stop: Callable[[], None]


async def process_check_ins_background(
    check_ins: list[City],
) -> ProcessingCheckIns:
    """チェックイン処理をバックグラウンドで開始し、制御ハンドルを返す。

    Examples:
        >>> import asyncio
        >>> async def example():
        ...     cities = [City("Tokyo"), City("Osaka"), City("Tokyo")]
        ...     processing = await process_check_ins_background(cities)
        ...     await asyncio.sleep(0.05)
        ...     ranking = await processing.current_ranking()
        ...     processing.stop()
        ...     return [s.city.name for s in ranking]
        >>> asyncio.run(example())
        ['Tokyo', 'Osaka']
    """
    stored_check_ins: AsyncRef[dict[City, int]] = AsyncRef({})
    stored_ranking: AsyncRef[list[CityStats]] = AsyncRef([])

    async def check_in_processor() -> None:
        for city in check_ins:
            await store_check_in(stored_check_ins, city)
            await asyncio.sleep(0)  # yield control

    async def ranking_updater() -> None:
        while True:
            await update_ranking(stored_check_ins, stored_ranking)
            await asyncio.sleep(0.01)

    check_in_fiber = start_fiber(check_in_processor())
    ranking_fiber = start_fiber(ranking_updater())

    def stop() -> None:
        check_in_fiber.cancel()
        ranking_fiber.cancel()

    async def get_ranking() -> list[CityStats]:
        return await stored_ranking.get()

    return ProcessingCheckIns(current_ranking=get_ranking, stop=stop)


# =============================================================================
# 10.11 並列 vs 逐次の比較
# =============================================================================


async def sequential_sleep(count: int, seconds: float) -> float:
    """逐次的にスリープする（合計時間を返す）。

    Examples:
        >>> import asyncio
        >>> import time
        >>> start = time.time()
        >>> asyncio.run(sequential_sleep(3, 0.1))
        >>> elapsed = time.time() - start
        >>> 0.3 <= elapsed < 0.5
        True
    """
    for _ in range(count):
        await asyncio.sleep(seconds)
    return count * seconds


async def parallel_sleep(count: int, seconds: float) -> float:
    """並列にスリープする（合計時間を返す）。

    Examples:
        >>> import asyncio
        >>> import time
        >>> start = time.time()
        >>> asyncio.run(parallel_sleep(3, 0.1))
        >>> elapsed = time.time() - start
        >>> elapsed < 0.2  # 並列なので約0.1秒
        True
    """
    await par_sequence([asyncio.sleep(seconds) for _ in range(count)])
    return seconds  # 並列なので最長のタスク時間のみ


# =============================================================================
# スレッドベースの並行処理
# =============================================================================


def run_in_threads(tasks: list[Callable[[], T]]) -> list[T]:
    """複数のタスクをスレッドで並列実行する。

    Examples:
        >>> def task1():
        ...     return 1
        >>> def task2():
        ...     return 2
        >>> sorted(run_in_threads([task1, task2]))
        [1, 2]
    """
    results: list[T] = []
    lock = threading.Lock()

    def run_task(task: Callable[[], T]) -> None:
        result = task()
        with lock:
            results.append(result)

    threads = [threading.Thread(target=run_task, args=(task,)) for task in tasks]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    return results


def parallel_map(items: list[T], f: Callable[[T], U]) -> list[U]:
    """リストの各要素に関数を並列適用する（スレッド使用）。

    Examples:
        >>> parallel_map([1, 2, 3], lambda x: x * 2)
        [2, 4, 6]
    """
    results: list[tuple[int, U]] = []
    lock = threading.Lock()

    def process(index: int, item: T) -> None:
        result = f(item)
        with lock:
            results.append((index, result))

    threads = [
        threading.Thread(target=process, args=(i, item))
        for i, item in enumerate(items)
    ]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    # インデックスでソートして結果を返す
    return [r for _, r in sorted(results)]

File: app/ruby/src/ch10_concurrency.rb
# frozen_string_literal: true

require_relative 'ch08_io'

# Chapter 10: Concurrency - Parallel and Concurrent Processing
# 並行処理 - 並列・並行プログラミング
module Ch10Concurrency
  # ===========================================================================
  # 10.1 Ref - Atomic Shared State
  # ===========================================================================

  # Ref provides thread-safe atomic operations on a value
  # Ref はスレッドセーフなアトミック操作を提供
  class Ref
    def initialize(initial_value)
      @value = initial_value
      @mutex = Mutex.new
    end

    # Create a new Ref with initial value (returns IO)
    # 初期値で新しい Ref を作成（IO を返す）
    def self.of(initial_value)
      Ch08IO::IO.delay { new(initial_value) }
    end

    # Get current value (returns IO)
    # 現在の値を取得（IO を返す）
    def get
      Ch08IO::IO.delay { @mutex.synchronize { @value } }
    end

    # Set value (returns IO)
    # 値を設定（IO を返す）
    def set(new_value)
      Ch08IO::IO.delay { @mutex.synchronize { @value = new_value } }
    end

    # Update value atomically (returns IO)
    # 値をアトミックに更新（IO を返す）
    def update(&fn)
      Ch08IO::IO.delay do
        @mutex.synchronize { @value = fn.call(@value) }
      end
    end

    # Modify and return old value (returns IO)
    # 変更して古い値を返す（IO を返す）
    def modify(&fn)
      Ch08IO::IO.delay do
        @mutex.synchronize do
          old_value = @value
          new_value, result = fn.call(old_value)
          @value = new_value
          result
        end
      end
    end

    # Update and get new value (returns IO)
    # 更新して新しい値を取得（IO を返す）
    def update_and_get(&fn)
      Ch08IO::IO.delay do
        @mutex.synchronize do
          @value = fn.call(@value)
          @value
        end
      end
    end

    # Get and update (returns IO with old value)
    # 取得して更新（古い値を返す IO）
    def get_and_update(&fn)
      Ch08IO::IO.delay do
        @mutex.synchronize do
          old_value = @value
          @value = fn.call(@value)
          old_value
        end
      end
    end
  end

  # ===========================================================================
  # 10.2 Parallel Execution
  # ===========================================================================

  # Execute IOs in parallel and collect results
  # IO を並列実行して結果を収集
  def self.par_sequence(ios)
    Ch08IO::IO.delay do
      threads = ios.map do |io|
        Thread.new { io.run! }
      end
      threads.map(&:value)
    end
  end

  # Map over array in parallel
  # 配列を並列にマップ
  def self.par_traverse(array, &fn)
    par_sequence(array.map(&fn))
  end

  # Execute two IOs in parallel and combine results
  # 2つの IO を並列実行して結果を結合
  def self.par_map2(io1, io2, &fn)
    Ch08IO::IO.delay do
      t1 = Thread.new { io1.run! }
      t2 = Thread.new { io2.run! }
      fn.call(t1.value, t2.value)
    end
  end

  # Execute two IOs in parallel and return both results
  # 2つの IO を並列実行して両方の結果を返す
  def self.par_both(io1, io2)
    par_map2(io1, io2) { |a, b| [a, b] }
  end

  # Race two IOs, return the first to complete
  # 2つの IO を競争させ、最初に完了したものを返す
  def self.race(io1, io2)
    Ch08IO::IO.delay do
      result = nil
      done = false
      mutex = Mutex.new
      cv = ConditionVariable.new

      [io1, io2].each do |io|
        Thread.new do
          value = io.run!
          mutex.synchronize do
            unless done
              result = value
              done = true
              cv.signal
            end
          end
        end
      end

      mutex.synchronize do
        cv.wait(mutex) until done
        result
      end
    end
  end

  # ===========================================================================
  # 10.3 Fiber - Lightweight Thread Simulation
  # ===========================================================================

  # FiberHandle represents a running background computation
  # FiberHandle はバックグラウンドで実行中の計算を表す
  class FiberHandle
    attr_reader :thread

    def initialize(thread, cancel_flag)
      @thread = thread
      @cancel_flag = cancel_flag
    end

    # Wait for fiber to complete and get result
    # Fiber の完了を待ち、結果を取得
    def join
      Ch08IO::IO.delay { @thread.value }
    end

    # Cancel the fiber
    # Fiber をキャンセル
    def cancel
      Ch08IO::IO.delay do
        @cancel_flag[:cancelled] = true
        @thread.kill if @thread.alive?
      end
    end

    # Check if fiber is still running
    # Fiber がまだ実行中か確認
    def alive?
      @thread.alive?
    end
  end

  # Start an IO as a background fiber
  # IO をバックグラウンド Fiber として開始
  def self.start(io)
    Ch08IO::IO.delay do
      cancel_flag = { cancelled: false }
      thread = Thread.new do
        io.run! unless cancel_flag[:cancelled]
      end
      FiberHandle.new(thread, cancel_flag)
    end
  end

  # ===========================================================================
  # 10.4 Forever and Repeat
  # ===========================================================================

  # Run an IO forever (until cancelled)
  # IO を永遠に実行（キャンセルされるまで）
  def self.forever_m(io)
    Ch08IO::IO.delay do
      loop do
        io.run!
      end
    end
  end

  # Run an IO n times
  # IO を n 回実行
  def self.repeat_n(io, n)
    Ch08IO::IO.delay do
      n.times.map { io.run! }
    end
  end

  # ===========================================================================
  # 10.5 Sleep
  # ===========================================================================

  # Sleep for specified seconds (non-blocking for other threads)
  # 指定秒数スリープ（他のスレッドをブロックしない）
  def self.sleep_io(seconds)
    Ch08IO::IO.delay { sleep(seconds) }
  end

  # Sleep for specified milliseconds
  # 指定ミリ秒スリープ
  def self.sleep_millis(millis)
    sleep_io(millis / 1000.0)
  end

  # ===========================================================================
  # 10.6 Check-ins Example
  # ===========================================================================

  City = Struct.new(:name, keyword_init: true)
  CityStats = Struct.new(:city, :check_ins, keyword_init: true)

  # Calculate top N cities from check-in counts (pure function)
  # チェックイン数からトップ N 都市を計算（純粋関数）
  def self.top_cities(city_check_ins, n = 3)
    city_check_ins
      .map { |city, check_ins| CityStats.new(city: city, check_ins: check_ins) }
      .sort_by { |stats| -stats.check_ins }
      .take(n)
  end

  # Store a check-in (updates Ref)
  # チェックインを保存（Ref を更新）
  def self.store_check_in(stored_check_ins, city)
    stored_check_ins.update do |check_ins|
      current = check_ins[city] || 0
      check_ins.merge(city => current + 1)
    end
  end

  # Update ranking from check-ins (pure function wrapped in IO)
  # チェックインからランキングを更新（IO でラップした純粋関数）
  def self.update_ranking(stored_check_ins, stored_ranking)
    stored_check_ins.get.bind do |check_ins|
      stored_ranking.set(top_cities(check_ins))
    end
  end

  # Process check-ins sequentially
  # チェックインを逐次処理
  def self.process_check_ins_sequential(check_ins, stored_check_ins)
    Ch08IO::IO.delay do
      check_ins.each do |city|
        store_check_in(stored_check_ins, city).run!
      end
    end
  end

  # ===========================================================================
  # 10.7 Concurrent Check-in Processing
  # ===========================================================================

  # Result of processing check-ins
  # チェックイン処理の結果
  ProcessingResult = Struct.new(:current_ranking, :stop, keyword_init: true)

  # Process check-ins concurrently with ranking updates
  # ランキング更新と並行してチェックインを処理
  def self.process_check_ins_concurrent(check_ins)
    Ch08IO::IO.delay do
      stored_check_ins = Ref.new({})
      stored_ranking = Ref.new([])

      cancel_flag = { cancelled: false }

      # Ranking update thread (runs continuously)
      ranking_thread = Thread.new do
        until cancel_flag[:cancelled]
          update_ranking(stored_check_ins, stored_ranking).run!
          sleep(0.01) # Small delay to prevent busy-waiting
        end
      end

      # Check-in processing thread
      checkin_thread = Thread.new do
        check_ins.each do |city|
          break if cancel_flag[:cancelled]

          store_check_in(stored_check_ins, city).run!
        end
      end

      # Wait for check-ins to complete
      checkin_thread.join

      # Allow final ranking update
      sleep(0.05)

      # Signal cancellation
      cancel_flag[:cancelled] = true
      ranking_thread.join

      stored_ranking.get.run!
    end
  end

  # Start processing and return handle for async control
  # 処理を開始し、非同期制御用のハンドルを返す
  def self.start_processing(check_ins)
    Ch08IO::IO.delay do
      stored_check_ins = Ref.new({})
      stored_ranking = Ref.new([])

      cancel_flag = { cancelled: false }

      # Start ranking update thread
      ranking_thread = Thread.new do
        until cancel_flag[:cancelled]
          update_ranking(stored_check_ins, stored_ranking).run!
          sleep(0.01)
        end
      end

      # Start check-in processing thread
      checkin_thread = Thread.new do
        check_ins.each do |city|
          break if cancel_flag[:cancelled]

          store_check_in(stored_check_ins, city).run!
        end
      end

      # Return control interface
      current_ranking = Ch08IO::IO.delay { stored_ranking.get.run! }
      stop = Ch08IO::IO.delay do
        cancel_flag[:cancelled] = true
        checkin_thread.kill if checkin_thread.alive?
        ranking_thread.join
      end

      ProcessingResult.new(current_ranking: current_ranking, stop: stop)
    end
  end

  # ===========================================================================
  # 10.8 Parallel Die Casting
  # ===========================================================================

  # Cast die twice in parallel and sum
  # サイコロを並列に2回振って合計
  def self.cast_die_twice_parallel
    par_map2(
      Ch08IO.cast_the_die,
      Ch08IO.cast_the_die
    ) { |a, b| a + b }
  end

  # Cast die n times in parallel
  # サイコロを並列に n 回振る
  def self.cast_die_n_parallel(n)
    ios = Array.new(n) { Ch08IO.cast_the_die }
    par_sequence(ios)
  end

  # Cast die n times in parallel and store in Ref
  # サイコロを並列に n 回振って Ref に保存
  def self.cast_die_n_with_ref(n)
    Ref.of([]).bind do |stored_casts|
      single_cast = Ch08IO.cast_the_die.bind do |result|
        stored_casts.update { |casts| casts + [result] }
      end
      ios = Array.new(n) { single_cast }
      par_sequence(ios).bind { stored_casts.get }
    end
  end

  # ===========================================================================
  # 10.9 Timeout Operations
  # ===========================================================================

  # Run IO with timeout
  # タイムアウト付きで IO を実行
  def self.with_timeout(io, timeout_seconds)
    Ch08IO::IO.delay do
      result = nil
      completed = false

      thread = Thread.new do
        result = io.run!
        completed = true
      end

      thread.join(timeout_seconds)

      if completed
        { success: true, value: result }
      else
        thread.kill
        { success: false, error: 'Timeout' }
      end
    end
  end

  # Collect results for a duration then stop
  # 一定時間結果を収集して停止
  def self.collect_for(producer_io, duration_seconds)
    Ch08IO::IO.delay do
      collected = Ref.new([])
      cancel_flag = { cancelled: false }

      producer_thread = Thread.new do
        until cancel_flag[:cancelled]
          result = producer_io.run!
          collected.update { |arr| arr + [result] }.run!
        end
      end

      sleep(duration_seconds)
      cancel_flag[:cancelled] = true
      producer_thread.kill if producer_thread.alive?

      collected.get.run!
    end
  end

  # ===========================================================================
  # 10.10 Parallel Map Updates
  # ===========================================================================

  Update = Struct.new(:key, :value, keyword_init: true)

  # Apply updates to a map in parallel
  # マップに並列で更新を適用
  def self.apply_updates(updates)
    Ref.of({}).bind do |map_ref|
      ios = updates.map do |update|
        map_ref.update { |m| m.merge(update.key => update.value) }
      end
      par_sequence(ios).bind { map_ref.get }
    end
  end

  # ===========================================================================
  # 10.11 Counter Examples
  # ===========================================================================

  # Increment counter n times sequentially
  # カウンターを順次 n 回インクリメント
  def self.increment_sequential(n)
    Ref.of(0).bind do |counter|
      n.times { counter.update { |x| x + 1 }.run! }
      counter.get
    end
  end

  # Increment counter n times in parallel
  # カウンターを並列に n 回インクリメント
  def self.increment_parallel(n)
    Ref.of(0).bind do |counter|
      ios = Array.new(n) { counter.update { |x| x + 1 } }
      par_sequence(ios).bind { counter.get }
    end
  end

  # Count evens from parallel IO executions
  # 並列 IO 実行から偶数をカウント
  def self.count_evens(ios)
    Ref.of(0).bind do |counter|
      check_ios = ios.map do |io|
        io.bind do |n|
          if n.even?
            counter.update { |c| c + 1 }
          else
            Ch08IO::IO.unit
          end
        end
      end
      par_sequence(check_ios).bind { counter.get }
    end
  end

  # ===========================================================================
  # 10.12 Producer-Consumer Pattern
  # ===========================================================================

  # Simple queue using Ref
  # Ref を使用したシンプルなキュー
  class RefQueue
    def initialize
      @ref = Ref.new([])
    end

    def enqueue(value)
      @ref.update { |arr| arr + [value] }
    end

    def dequeue
      @ref.modify do |arr|
        if arr.empty?
          [arr, nil]
        else
          [arr[1..], arr[0]]
        end
      end
    end

    def size
      @ref.get.fmap(&:size)
    end

    def to_array
      @ref.get
    end
  end

  # Create a producer-consumer pair
  # プロデューサー・コンシューマーのペアを作成
  def self.producer_consumer(producer_io, consumer_fn, duration_seconds)
    Ch08IO::IO.delay do
      queue = RefQueue.new
      results = Ref.new([])
      cancel_flag = { cancelled: false }

      # Producer thread
      producer_thread = Thread.new do
        until cancel_flag[:cancelled]
          value = producer_io.run!
          queue.enqueue(value).run!
        end
      end

      # Consumer thread
      consumer_thread = Thread.new do
        until cancel_flag[:cancelled]
          value = queue.dequeue.run!
          if value
            result = consumer_fn.call(value)
            results.update { |arr| arr + [result] }.run!
          else
            sleep(0.001) # Wait for producer
          end
        end
      end

      sleep(duration_seconds)
      cancel_flag[:cancelled] = true
      producer_thread.kill if producer_thread.alive?
      consumer_thread.kill if consumer_thread.alive?

      results.get.run!
    end
  end
end
